File: rust/3.longest-substring-without-repeating-characters.rs
/*
 * @lc app=leetcode id=3 lang=rust
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut ascii: [usize; 128] = [0; 128];
        let mut length: i32 = 0;
        let mut start: usize = 0;
        for (i, c) in s.chars().enumerate() {
            if ascii[c as usize] >= start {
                length = length.max((i - start) as i32);
                start = ascii[c as usize];
            }
            ascii[c as usize] = i + 1;
        }
        return length.max((s.len() - start) as i32);
    }
}
// @lc code=end

File: javascript/3.longest-substring-without-repeating-characters.js
/*
 * @lc app=leetcode id=3 lang=javascript
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
  if (s.length <= 1) return s.length;
  const chars = new Array(128).fill(0);
  let slow = 0,
    fast = 0,
    length = 0;
  for (; fast < s.length; fast++) {
    const c = s[fast].charCodeAt();
    if (chars[c] >= slow) {
      length = Math.max(fast - slow, length);
      slow = chars[c];
    }
    chars[c] = fast + 1;
  }
  return Math.max(fast - slow, length);
};
// @lc code=end
